### PR TITLE
Added Gallery Model Class

### LINKED ISSUE #32: Design Gallery Model class

### PR DESCRIPTION
**Pull request by shikdernyc**
## Summary
- Added Gallery Model class
- Added `GetUserGallery()` method to retrieve a user's gallery
- Added `loadMore()` method to fetch more photos and handle pagination
- Added `ResultCallback` interface which acts as a generic callback function that contains an onSuccess and onError callback
- Refactored folder structures using the groups `activities`, `interfaces`, `models`, and `services`
- Resolves #32

### Retrieve Gallery for a User
```
UserGallery.GetUserGallery("8b8b14a9-b180-4116-a73c-b6dc5a5fb291", new ResultCallback<UserGallery, IOException>() {
    @Override
    public void onSuccess(UserGallery gallery) {
        Log.i(tag, gallery.toString());
     }

    @Override
    public void onError(IOException error) {
        Log.e(tag, error.getMessage());
    }
});
```
In the example above, we are retrieving user gallery for user with id `8b8b14a9-b180-4116-a73c-b6dc5a5fb291`. The onSuccess method's `gallery` parameter contains the user gallery. You can view all the entries in the gallery by running 
`gallery.toString()`

##### Output of toString() method
```
2021-04-26 16:45:21.401 24340-24412/com.example.mealsnap_prototype_2 I/DummyFN: UserGallery{
    	Total items=10, nextPage='/user/8b8b14a9-b180-4116-a73c-b6dc5a5fb291/gallery?page_pk=userid%238b8b14a9-b180-4116-a73c-b6dc5a5fb291&page_sk=photoId%23f6547e93-f146-4664-9dc0-b3652c715450
    
    	GalleryPhoto{id='4c0239eb-2cdd-4d95-b04e-cabd06be70e5', title='F', description='F', imageUrl='https://mealsnapappstack-prod-mealsnappostimagestoragepro-2hts76vt6mq5.s3.amazonaws.com/dd84a7bb-5edc-4bdd-a1fa-1fd0f0cd565c.jpg', alreadyLiked=false, likeCount=0}
    	GalleryPhoto{id='6633624b-6b67-4d2f-b532-2e3abbf9d5e4', title='E', description='E', imageUrl='https://mealsnapappstack-prod-mealsnappostimagestoragepro-2hts76vt6mq5.s3.amazonaws.com/a04e3fa4-4f9f-4768-b3a6-9958e3d75219.jpg', alreadyLiked=false, likeCount=0}
    	GalleryPhoto{id='73770187-e933-4cbb-b686-1648c38ea592', title='Assad', description='ASd', imageUrl='https://mealsnapappstack-prod-mealsnappostimagestoragepro-2hts76vt6mq5.s3.amazonaws.com/33916101-6875-455e-91ec-6bf33dcc6d3b.jpg', alreadyLiked=false, likeCount=0}
    	GalleryPhoto{id='78ceb7b5-fbb2-4886-8530-5a05586a936a', title='C', description='C', imageUrl='https://mealsnapappstack-prod-mealsnappostimagestoragepro-2hts76vt6mq5.s3.amazonaws.com/b8c15a63-e642-4bd4-ba30-9ca048e72a23.jpg', alreadyLiked=false, likeCount=0}
    	GalleryPhoto{id='798d92e9-f550-4d1f-b752-7c4379c6dfae', title='ABC', description='ABC', imageUrl='https://mealsnapappstack-prod-mealsnappostimagestoragepro-2hts76vt6mq5.s3.amazonaws.com/faeddb7d-1c09-48fb-8757-cd1e2e94e8b2.jpg', alreadyLiked=false, likeCount=0}
    	GalleryPhoto{id='8b2b0570-3d40-46bb-a29d-b7a4450ac485', title='B', description='B', imageUrl='https://mealsnapappstack-prod-mealsnappostimagestoragepro-2hts76vt6mq5.s3.amazonaws.com/da5e40b3-6869-4042-b8c4-4a5ca194fe26.jpg', alreadyLiked=false, likeCount=0}
    	GalleryPhoto{id='97abebc6-bb48-4908-82fc-68d97e89cc0c', title='D', description='D', imageUrl='https://mealsnapappstack-prod-mealsnappostimagestoragepro-2hts76vt6mq5.s3.amazonaws.com/951f4b58-ffaf-414a-b927-a807e22ba058.jpg', alreadyLiked=false, likeCount=0}
    	GalleryPhoto{id='9f9d724c-ace9-417b-9725-f1bccd864a71', title='asd', description='Asd', imageUrl='https://mealsnapappstack-prod-mealsnappostimagestoragepro-2hts76vt6mq5.s3.amazonaws.com/1546334f-ea9d-4461-8d01-63b4ca39df64.jpg', alreadyLiked=false, likeCount=0}
    	GalleryPhoto{id='bd8f8057-0b80-400d-98b8-1b5a8366e55f', title='ASDasd', description='ASd', imageUrl='https://mealsnapappstack-prod-mealsnappostimagestoragepro-2hts76vt6mq5.s3.amazonaws.com/eb6cb8ab-6d37-493e-904f-2aa8cb2a171b.jpg', alreadyLiked=false, likeCount=0}
    	GalleryPhoto{id='f6547e93-f146-4664-9dc0-b3652c715450', title='H', description='H', imageUrl='https://mealsnapappstack-prod-mealsnappostimagestoragepro-2hts76vt6mq5.s3.amazonaws.com/d9551874-d601-41ab-825a-5473805a4eb0.jpg', alreadyLiked=false, likeCount=0}
    }
```

### Viewing currently fetched items in a `UserGallery`
You can view all the items in a user gallery by running `items()` method. This returns a `List<GalleryPhoto`.

### Loading More Photos
User Gallery does not contain all the user photos but rather are retrieved in chunks. You can fetch the next chunk by running the `loadMore()` function such as the following
```
gallery.loadMore(new ResultCallback<Boolean, Exception>() {
    @Override
    public void onSuccess(Boolean _) {
        // More image has been loaded
        Log.i(tag, gallery.toString());
    }

    @Override
    public void onError(Exception error) {
        if(error instanceof EOFException){
            Log.i(tag, "No more pages");
        }else{
            Log.i(tag, error.getMessage());
        }
    }
});
```
In the above snippet, onSuccess indicates more images have been added to the gallery list. There would be a point when all images have been fetched. Running loadMore, in that case, would cause an EOFException as shown by the `onError()` function.

### CODE REVIEW

File: app/src/main/java/com/example/mealsnap_prototype_2/interfaces/ResultCallback.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package com.example.mealsnap_prototype_2.interfaces;
```
Comment (shikdernyc) [inline app/src/main/java/com/example/mealsnap_prototype_2/interfaces/ResultCallback.java:1]: New Interface

File: app/src/main/java/com/example/mealsnap_prototype_2/models/gallery/FetchUserPhotoAPIResponse.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package com.example.mealsnap_prototype_2.models.gallery;
+
+import com.example.mealsnap_prototype_2.services.appapi.common.PageResponse;
+
+class FetchUserPhotoAPIResponse {
```
Comment (shikdernyc) [inline app/src/main/java/com/example/mealsnap_prototype_2/models/gallery/FetchUserPhotoAPIResponse.java:5]: Newly Added

File: app/src/main/java/com/example/mealsnap_prototype_2/models/gallery/GalleryPhoto.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
+package com.example.mealsnap_prototype_2.models.gallery;
+
+class GalleryPhoto {
```
Comment (shikdernyc) [inline app/src/main/java/com/example/mealsnap_prototype_2/models/gallery/GalleryPhoto.java:3]: Newly Added

File: app/src/main/java/com/example/mealsnap_prototype_2/models/gallery/UserGallery.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+import java.io.EOFException;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+
+import okhttp3.ResponseBody;
+import com.example.mealsnap_prototype_2.services.appapi.APIRequest;
+
+public class UserGallery {
```
Comment (shikdernyc) [inline app/src/main/java/com/example/mealsnap_prototype_2/models/gallery/UserGallery.java:19]: Newly Added

File: app/src/main/java/com/example/mealsnap_prototype_2/services/appapi/common/PageResponse.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package com.example.mealsnap_prototype_2.services.appapi.common;
+
+import androidx.annotation.Nullable;
+
+// Reusable response data class for requests that returns page response
+// Example Response: https://www.notion.so/API-Documentation-c622f4f78c7941c987a4362acb73b6cc#f1ebaa442a3d4f0a81291cba55cb6a10
+public class PageResponse {
```
Comment (shikdernyc) [inline app/src/main/java/com/example/mealsnap_prototype_2/services/appapi/common/PageResponse.java:7]: Newly Added

File: app/src/main/java/com/example/mealsnap_prototype_2/services/appapi/APIRequest.java [hunk 6]
```diff
@@ -21,22 +18,35 @@
...
     private static final OkHttpClient client = new OkHttpClient();
+    private static final String ROOT_URL = "https://dbkw974ay1.execute-api.us-east-1.amazonaws.com/prod/";
+
+    /**
+     * Generates appropriate request url given a route
+     *
+     * @param route Request Route. Ex: /echo | echo
+     * @return Reformatted URL
+     */
+    private static String GetRequestURL(String route) {
```
Comment (shikdernyc) [inline app/src/main/java/com/example/mealsnap_prototype_2/services/appapi/APIRequest.java:34]: @svhnstorage Newly Added

**Review by svhnstorage — APPROVED** [review]

**Review by qiaoganchen22 — APPROVED** [review]

**Pull request merged**